docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+limits the types of preprocessing that can be done, as the preprocessing must be done client-side and be constrained by
+the computational resources available to the client.
+
+In the planned distributed version of Chroma, the proxy layer
+that communicates with the client and the processing layer could theoretically be on different machines. This means
+that the processing layer could have access to more computational resources than the client. In addition, the processing
+could be done in parallel, which would allow for faster processing.
+
+5. Support for index customization and multiple indices
+
+Chroma currently implicitly creates two segments for every collection, one for the embeddings - the vector segment
+and one for the metadata - the metadata segment. In order to parameterize the index, users must pass in a set of
+parameters to the metadata of the collection itself, which is then used to create the index. For example
+
+```python
+collection = client.create_collection("my_collection", embedding_function=OpenAIEmbeddingFunction(), metadata={"hnsw:ef_construction": 200})
+```
+
+This is error-prone, as simple typos go unchecked and are interpreted silently as generic metadata, and also is not discoverable.
+Additionally, this means that users cannot create custom additional indices for their collections. For example, it may make sense
+that a user wants to not only control the parameters of the vector segment but also create a custom index for the metadata segment, create
+alternative vector segments with different parameters,
+
+6. Composition with Language Model Generation
+
+In the future, we may want to support the full RAG loop within Chroma. This means that we would want to support being able to take the output of retrieval and use it as input to a language model. This would require us to be able to compose the embedding function with a language model. This is not possible today, as the embedding function is not a first class citizen in Chroma.
+
+## Proposal
+
+We introduce three concepts in order to solve these problems - Pipelines, The Registry, and Storage
+
+### Pipelines
+
+A pipeline is a set of steps that are applied to the inputs to a chroma insert call. Broadly, everything including the segments written to are a part of the pipeline. The pipeline is responsible for loading the document, preprocessing the document, and writing the document to the segments. Pipeline steps are composable functions that can be chained together to form a pipeline. Each pipeline step is registered with a name, and can be referenced by that name. This registration can be thought of as similar to SQL functions.
+
+#### Pipeline Step Registration - The Registry
+
+For example, if you wanted to create a pipeline that chunked a document, you could do something like this:
+
+First, register The "Chunk" pipeline step with the name "Chunk" as follows:
+
+```python
+@pipeline_step(name="Chunk")
```

Comment:
If we use a chained syntax:

`withData(data).pipe([Chunk,Embed])` then we can even remove the decorator part - the actual wrapping can happen within the `pipe()`. 

Elaborating further on the chained pipeline:

- `.register(pipeline_name="my_pipeline")` or `.upload(pipeline_name="my_pipeline")` - Packages the pipeline and uploads to server
- `.exec_local()` - local execution for testing/dev